[1~3단계 - 방탈출 사용자 예약] 헤일러(민서현) 미션 제출합니다.

threepebbles opened the pull request:
안녕하세요 매트! 리뷰이 헤일러입니다.

페어와 저 모두 스프링에 익숙하지 않아서, 기능 구현을 완료하는 것을 목표로 페어 프로그래밍을 진행했습니다.

프롤로그는 조금 더 학습 후에 작성해서 올리려고 합니다!

부족한 부분이 많지만 리뷰 잘 부탁드립니다 😄

아래는 미션2 시작 이후 작성한 변경 사항 커밋 목록입니다.

[미션2 Files changed](https://github.com/woowacourse/spring-roomescape-member/pull/195/files/dbfdba8d33460c9aee2e64671e18828c028fd09a..4682df5058f85614382a3004e743e1751c910030)
## 체크 리스트

- [x]  미션의 필수 요구사항을 모두 구현했나요?
- [x]  Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x]  애플리케이션이 정상적으로 실행되나요?
- [x]  [프롤로그](https://prolog.techcourse.co.kr/)에 셀프 체크를 작성했나요?
<!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ]  1 (전혀 충족하지 못함)
- [ ]  2
- [x]  3 (보통)
- [ ]  4
- [ ]  5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

✅: 잘 지킴, ❓: 애매함

규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다. ✅
규칙 2: else 예약어를 쓰지 않는다. ✅
규칙 3: 모든 원시값과 문자열을 포장한다. ❓
규칙 4: 한 줄에 점을 하나만 찍는다. ✅
규칙 5: 줄여쓰지 않는다(축약 금지). ✅
규칙 6: 모든 엔티티를 작게 유지한다. ✅
규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다. ❓
규칙 8: 일급 콜렉션을 쓴다. ❓
규칙 9: 게터/세터/프로퍼티를 쓰지 않는다. ❓

## 웹 개발 경험 체크

🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?

- [ ]  아니오. 전혀 경험이 없습니다.
- [x]  예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ]  예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ]  예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크

- [x]  이전 미션의 내 코드에서 시작
- [ ]  이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?

이번 미션은 Service, Repository, Domain 계층 간의 의존 관계, 그리고 admin 기능과 user 기능 간의 관계(권한 기능 확장 고려)에 대해 많은 고민을 해보게 되는 미션이었던 것 같습니다.

리뷰 잘 부탁드립니다 😄 

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)

✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.

✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!

✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.

코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.

✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

threepebbles committed:
migrate from mission 1

bowook committed:
Merge branch 'step1' of https://github.com/threepebbles/spring-roomescape-member into step1

bowook committed:
feat : 예약 시간 생성 API 응답 상태 코드 수정

- 200 -> 201

bowook committed:
feat : 예약 생성 API 응답 상태 코드 수정

- 200 -> 201

bowook committed:
refactor : 응답 메시지 시간 형식 지정

bowook committed:
refactor : 요청 메시지 시간 형식 지정

bowook committed:
docs : 발생할 수 있는 예외 상황 추가

bowook committed:
feat : 예약 시간 범위 검증 추가

bowook committed:
feat : 예약시간 중복 검증 추가

bowook committed:
feat : 삭제하려고 하는 예약 시간을 사용하고 있는 예약 정보가 있는지 검증하는 기능 추가

bowook committed:
chore : 테마 테이블 스키마 추가

bowook committed:
feat : Domain Entity와 Database Entity 분리

bowook committed:
chore : 테마 테이블명 변경

- theme -> themes

bowook committed:
feat : 테마 Domain Entity, Database Entity 추가

bowook committed:
feat : 테마 레포지토리 추가

bowook committed:
feat : 테마 관리 API 추가

- 테마 조회, 추가, 삭제 API

bowook committed:
refactor : 메서드명 수정

bowook committed:
feat : 참조 중인 예약 정보 행을 삭제 요청시 발생하는 예외 처리 수정

bowook committed:
refactor : 메서드명 수정

threepebbles committed:
docs : 테마 요구사항 추가

threepebbles committed:
feat : 중복된 테마명 검증 기능 추가

threepebbles committed:
chore : URL 슬래시 추가

threepebbles committed:
feat : 한 테마에 대해서 같은 날짜와 시간에 예약된 정보가 있는지 검증 추가

threepebbles committed:
docs : 사용자 예약 화면 응답 추가

threepebbles committed:
feat : 사용자 예약 화면 응답 기능 추가

threepebbles committed:
chore : 생성자 검증 로직 리팩터링

threepebbles committed:
chore : 테마명에 UNIQUE 제약 조건 추가

threepebbles committed:
feat : 날짜와 테마명에 따른 예약 가능한 시간 검증 기능 추가

threepebbles committed:
feat : 예약 시간 제거 API 응답 상태 코드 수정

threepebbles committed:
chore : 불필요한 공백 제거

threepebbles committed:
chore : final 키워드 추가

threepebbles committed:
refactor : Domain Entity, Database Entity 병합

threepebbles committed:
docs : 유저 예약 시스템 추가

threepebbles committed:
feat : 홈 화면 API 응답 추가

threepebbles committed:
feat : 일주일 이내 가장 예약이 많았던 인기 테마 10개를 응답하는 API 추가

bowook committed:
chore : 메서드 선언 순서 변경

bowook committed:
test : 예약 서비스 테스트 추가

bowook committed:
test : 예약 생성자 검증 테스트 추가

threepebbles committed:
test : 날짜와 시간이 중복되는 테마 검증 테스트 추가

threepebbles committed:
test : 예약 시간 테스트 추가

threepebbles committed:
test : 테마 서비스 테스트 추가

threepebbles committed:
test : 테마 테스트 추가

threepebbles committed:
chore : 공백 정리

threepebbles committed:
test : 예약 시간 컨트롤러 API 테스트 추가

threepebbles committed:
test : 테마 컨트롤러 API 테스트 추가

threepebbles committed:
test : TestConfiguration 추가

threepebbles committed:
feat : 예약 정보 삭제 API 응답 상태 코드 수정

- ok -> no content

threepebbles committed:
test : 예약 컨트롤러 API 테스트 추가

bowook committed:
chore : 공백 정리

bowook committed:
chore : 변수명 오타 수정

bowook committed:
chore : 패키지 이동

bowook committed:
refactor : SQL Query 개선

threepebbles committed:
remove : 테스트 schema.sql 파일 삭제

hyeonic commented on build.gradle:
```diff
@@ -17,13 +17,23 @@ repositories {
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
 
+    // jdbc
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
```
중복된 의존성이 들어간 것 같네요!

threepebbles replied:
감사합니다 수정했습니다!

반영 커밋: 8863987928fd5d2b01ca39b20da820eed795f07a

hyeonic commented on build.gradle:
```diff
@@ -17,13 +17,23 @@ repositories {
+    // jdbc
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+
+    // Lombok
+    annotationProcessor 'org.projectlombok:lombok'
+    compileOnly 'org.projectlombok:lombok'
```
어떠한 목적으로 롬복 라이브러리를 사용하였나요??

threepebbles replied:
Lombok은 Java로 개발할 때 자주 반복적으로 작성되는 보일러 플레이트 코드(ex. getter/setter, 생성자, EqualsAndHashCode 등)를 컴파일의 어노테이션 프로세싱 단계에서 생성하기 위해 만들어진 라이브러리입니다.

Lombok 관련 어노테이션을 사용하면 getter, setter 등의 다소 중요하지 않은 기능들을 감추어 클래스 핵심 로직에 대한 가독성을 향상시켜주고, 작성해야 하는 절대적인 소스 코드의 양이 적어져 개발 생산성이 크게 향상된다고 생각하여 사용했습니다.

단, Lombok 관련 어노테이션으로 생성되는 소스 코드가, 실제 자바 소스코드로 표현됐을 때 어떻게 구현되는지 명확하게 알고 사용해야 한다고 생각합니다. 😃

hyeonic replied:
>  단, Lombok 관련 어노테이션으로 생성되는 소스 코드가, 실제 자바 소스코드로 표현됐을 때 어떻게 구현되는지 명확하게 알고 사용해야 한다고 생각합니다. 😃

좋은 마음가짐인 것 같아요 ㅎㅎ 롬복이 아니더라도 어떠한 라이브러리를 도입할 때 내부의 동작 구조를 모른다면 작성한 코드를 온전히 신뢰하긴 어려울 것 같아요. 이후 고려하실 때도 헤일러만의 충분한 근거가 있으면 좋을 것 같습니다!

hyeonic commented on src/main/java/roomescape/exception/DataExistException.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.exception;
+
+public class DataExistException extends RuntimeException {
+
+    public DataExistException(String message) {
+        super(message);
+    }
+}
```
1. 커스텀한 예외를 구성한 이유는 무엇인가요?
2. RuntimeException은 어떠한 특징을 가지고 있나요?

threepebbles replied:
> 커스텀한 예외를 구성한 이유는 무엇인가요?

제가 커스텀 예외를 도입하게 된 이유는 비즈니스 로직 상 자주 발생할 수 있는 예외 상황에 대해 일관되게 예외 처리를 하고 싶어서였습니다. 

그리고 Java에서 기본으로 제공해주는 Exception만으로는 데이터 조회 실패와 같은 예외의 상황을 명확하게 드러내기 어려울 것 같아서 커스텀 예외를 사용해보았습니다.

또한 JDBC에서 사용하는 예외를 사용하는 방법이 있는데 예외 로직이 JDBC라는 특정 구현체에 의존하는 것이 마음에 들지 않아(?) 사용하지 않았습니다.

> RuntimeException은 어떠한 특징을 가지고 있나요?

RuntimeException은 컴파일러가 예외 처리를 강제하지 않는 예외(복구 불가능한 예외, Unchecked Exception)임을 의미합니다. 복구 불가능한 예외에 대한 예를 들면, 데이터베이스 테이블에 존재하지 않는 id에 접근하려는 행위가 그렇습니다. 프로그램 입장에서는 존재하지 않는 id에 접근하려는 행위를 중단시키기만 할 수 있을 뿐, 재시도 혹은 복구가 불가능한 로직입니다. `존재하지 않는 데이터 조회 시도`, `이미 존재하는 테마 이름으로 테마 생성 시도` 과 같은 대부분의 비즈니스 로직 상의 예외는 복구 불가능한 예외이므로 RuntimeException에 해당된다고 생각했습니다!

hyeonic replied:
넵 좋습니다! `Unchecked Exception`과 더불어서 `Checked Exception`에 대한 특성과 왜 등장 했는지, 어디에서 활용할 수 있는지 고민해보면 좋을 것 같아요 ㅎㅎ

hyeonic commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
`@ControllerAdvice`의 목적은 무엇인가요? `@RestControllerAdvice`와는 어떻게 다른가요?

threepebbles replied:
`@RestControllerAdvice` 에 대해서는 존재를 모르고 있어서 학습해보았는데, `@Contoller` 와 `@RestContoller` 의 차이와 동일한 것 같습니다!

`@RestControllerAdvice` 내부를 보면 단순히 `@ControllerAdvice` 와 `@ResponseBody` 의 조합입니다.

```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@ControllerAdvice
@ResponseBody
public @interface RestControllerAdvice {
}
```

`@ControllerAdvice` 는 템플릿 엔진을 이용한 뷰 응답과 JSON 응답이 모두 가능하고, `@RestControllerAdvice` 는 JSON 응답만 가능합니다.

현재 GlobalExceptionHandler에서 처리하고 있는 예외들은 REST API에서 발생한 예외들이고, 응답의 body가 모두 JSON 형식인 경우기 때문에, `@RestControllerAdvice`를 사용하는 것이 좀 더 명확한 역할을 드러낼 수 있을 것 같습니다!

반영 커밋: 9df51f1090ca77b2541e555ba394029ddf11d302

hyeonic commented on src/main/java/roomescape/reservation/repository/H2ReservationRepository.java:
```diff
@@ -0,0 +1,154 @@
+
+@Repository
+@RequiredArgsConstructor
+public class H2ReservationRepository implements ReservationRepository {
```
H2로 DB를 한정할 필요가 있을까요? 추후 MySQL을 고려할 수도 있는데 Jdbc 기반으로 RDB를 사용한다는 의미를 내포할 순 없을까요?

threepebbles replied:
H2ReservationRepository와 H2ReservationTimeRepository는 jdbc와 표준 SQL 문법에 기반하여 작성된 구현체기 때문에 JdbcReservationRepository와 JdbcReservationTimeRepository이 더 어울리는 네이밍 같습니다! 🙂

H2ThemeRepository에서는 H2에서만 동작하는 SQL 키워드(`DateADD`)를 사용하고 있는데 이와 같이 SQL 쿼리문이 H2에 종속적인 경우에는 H2 접두사를 붙이는 게 자연스러울까요?

hyeonic replied:
음.. 어디까지 구체적인 이름을 활용할지는 헤일러가 선택할 수 있을 것 같아요. 현재는 DB에 대한 구현체만 개발하였지만 이후에 애플리케이션의 규모가 커지고 다양한 인프라를 구성해야 할 때 지금과 유사한 고민을 할 수 있는데요, 저는 구현체더라도 추상화하여 표현할 수 있다면 최대한 반영하려 합니다. H2와 같이 매우매우 구체적인 이름이 추후 변경에 방해가 되기도 하더라구요 ㅎㅎ

요 부분은 고민해보시고 다음 단계 진행할 때 ThemeRepository도 같이 고려해주세요~ 일관성이 없는게 조금 어색하게 느껴지네요

hyeonic commented on src/main/java/roomescape/reservation/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Optional;
+import roomescape.reservation.domain.Reservation;
+
+public interface ReservationRepository {
+
+    Long save(Reservation reservation);
+
+    void deleteById(Long id);
+
+    boolean existsByDateAndStartAtAndThemeId(LocalDate date, LocalTime startAt, Long themeId);
+
+    boolean existsByDateAndTimeIdAndThemeId(LocalDate date, Long timeId, Long themeId);
+
+    Optional<Reservation> findById(Long id);
+
+    List<Reservation> findAll();
+}
```
이렇게 인터페이스를 구성하였을 때 어떠한 이점을 누릴 수 있을까요?

threepebbles replied:
Service 계층에서 인터페이스에 의존하게 하면,  추후 레포지토리 구현체가 변경되어도 Service 계층의 코드는 수정 없이 사용할 수 있다는 장점이 있습니다.

SOLID원칙에서 DIP, OCP를 잘 지킬 수 있는 구조라고 할 수 있습니다!

hyeonic replied:
좋습니다 ㅎㅎ 더 나아가 IoC, DI에 대한 개념도 자연스럽게 사용하며 습득해보면 좋을 것 같아요!

hyeonic replied:
좋습니다 ㅎㅎ 더 나아가 IoC, DI에 대한 개념도 자연스럽게 사용하며 습득해보면 좋을 것 같아요!

hyeonic commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,74 @@
+import roomescape.theme.domain.Theme;
+import roomescape.theme.repository.ThemeRepository;
+
+@Service
+@RequiredArgsConstructor
+public class ReservationService {
+
+    private final ReservationRepository reservationRepository;
+    private final ReservationTimeRepository reservationTimeRepository;
+    private final ThemeRepository themeRepository;
```
헤일러가 생각하는 domain, repository, service는 각각 무엇을 의미하나요?

threepebbles replied:
제가 생각하는 domain, service, repository는 다음과 같습니다.

우선 service와 domain은 서로 협력하여, 현실 세계의 요구 사항을 해결하기 위한 비즈니스 로직을 수행하는 역할을 합니다.

**domain 계층**에서는 **각 도메인 객체 별로 본인이 책임져야 하는 최대한 순수한 상태와 행위, 규칙들을 정의**합니다. 예를 들면 예약(Reservation)이라는 도메인 객체는 `예약 시간은 오전10시~오후10시 사이만 가능하다` 와 같은 제약 조건을 책임집니다.

**service 계층**은 **여러 도메인 객체들을 협력시켜, 유스케이스(Use Case)를 수행**하기 위한 계층입니다. 유스케이스는 `외부 클라이언트가 내부 시스템과 상호작용하여 달성하고자 하는 하나의 작업 단위` 를 의미합니다. 저는 service 계층에서 하나의 메서드가 하나의 유스케이스를 수행한다고 생각하고 작업하고 있습니다. 하나의 예를 들면 `date 날짜에 themeId에 해당하는 테마로 예약가능한 시간들을 조회하는 것` 이 외부 클라이언트의 요구 사항이고 이를 해결하기 위한 기능은 `findAvailableReservationTimes(final LocalDate date, final Long themeId)` 가 제공해주고 있습니다.

domain 계층에서 수행하는 로직을 service에서 수행할 수도 있지만, 모든 요구 사항에 대한 도메인 로직을 service에서 처리한다면 service 계층이 너무 비대해질 수 있습니다. 

service 계층이 비대해지면 응집도가 낮아지고, 재사용성도 낮아지고, 가독성도 떨어진다는 단점이 있기 때문에 service 계층은 최대한 실행의 흐름만 담당하도록 하고 domain 계층이 핵심적인 로직을 담당하게 하도록 노력하는 것이 좋다고 생각합니다.

**repository 계층**은 도메인 객체에 대한 CRUD 기능을 제공하는, **도메인 객체 저장소를 추상화한 계층**입니다. 도메인 객체를 인메모리에 일시적으로 저장하는 것, 파일 혹은 데이터베이스(ex. MySQL)에 영속적으로 저장하는 것 등은 저장 방식에 대한 세부 구현 사항이고, repository 계층의 책임은 저장 방식과 무관하게 도메인 객체의 CRUD 기능을 일관되게 제공해주는 것입니다.

hyeonic replied:
좋습니다! 헤일러가 생각하는 Repository에 대한 정의 덕분에 DB 모델링과 객체지향적인 패러다임 사이의 간극을 일정 부분 해소해 준 것으로 보여요 👍

hyeonic commented on src/main/java/roomescape/reservation/controller/ReservationRestController.java:
```diff
@@ -0,0 +1,73 @@
+import roomescape.reservation.domain.Reservation;
+import roomescape.reservation.service.ReservationService;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/reservations")
```
사소하지만 애노테이션이 늘어남에 따라 순서 배치에 대한 일관성을 지키는 것도 코드 파악에 도움이 되곤 합니다! 참고 부탁드려요 ㅎㅎ

threepebbles replied:
어노테이션 순서 배치에 대해서는 크게 생각해보지 못했었는데 짚어주셔서 감사합니다! 🙂

코드 가독성과 일관성을 위해 다음과 같이 어노테이션 순서를 정리했습니다!
주관적으로 중요하다고 생각되는 어노테이션을 위에 배치했습니다.
1. Spring Stereotype 어노테이션 (ex. Component, Controller, Service)
2. 생성자/필드 관련 어노테이션 (ex. Lombok 관련 어노테이션)
3. 핸들러 매핑 관련 어노테이션 (ex. RequestMapping)

반영 커밋: ca77b30ca49a5978aaabef19dac4d7221ef0ec4d

hyeonic commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,311 @@
+                .isInstanceOf(DataExistException.class);
+    }
+
+    @TestConfiguration
+    static class TestConfig {
+
+        @Bean
+        public ReservationTimeRepository reservationTimeRepository(
+                final JdbcTemplate jdbcTemplate
+        ) {
+            return new H2ReservationTimeRepository(jdbcTemplate);
+        }
+
+        @Bean
+        public ReservationRepository reservationRepository(
+                final JdbcTemplate jdbcTemplate
+        ) {
+            return new H2ReservationRepository(jdbcTemplate);
+        }
+
+        @Bean
+        public ReservationTimeService reservationTimeService(
+                final ReservationTimeRepository reservationTimeRepository
+        ) {
+            return new ReservationTimeService(reservationTimeRepository);
+        }
```
각 서비스마다 관련 설정을 구분한 이유가 있을까요??

threepebbles replied:
테스트 전역에 `@TestConfiguration`을 적용하지 않은 이유는, 각 Service 테스트에서 필요한 구현체를 독립적으로 구성하고 교체 가능한 구조를 한번 만들어보고 싶었기 때문입니다.
    
하나의 Service 클래스에 대해 MySQL, H2, Fake 레포지토리 구현체를 각각 주입하여 테스트하는 방식으로 벤치마킹 해볼 생각을 잠깐 했었습니다. 😅

각각에 대해 다른 TestConfig 클래스를 만들기에는 적절한 네이밍도 생각나지 않고, 해당 테스트 클래스에만 적용되는 설정이기 때문에 내부 클래스로 두게 되었습니다.
    
지금 상황에서는 TestConfig를 일반 클래스로 분리하고 `ReservationTimeServiceTest`를 아래와 같이 `@Import`를 사용하는 방식으로 수정해도 기능적으로 동일하게 동작할 것 같습니다.

```java
@JdbcTest
@Import(TestConfig.class)
class ReservationTimeServiceTest {
	...
}
```

hyeonic reviewed (changes requested):
안녕하세요 헤일러~ 방탈출 사용자 예약 리뷰를 맡은 매트입니다.

간단히 코멘트 남겼으니 확인해주시고 답변 남겨주세요! 질문이 모호하거나 추가적으로 궁금한 부분 있다면 DM 남겨주세요 ㅎㅎ

threepebbles committed:
chore: 중복된 의존성 제거

threepebbles committed:
refactor: 커스텀 예외 클래스 네이밍 개선

threepebbles committed:
refactor: 레포지토리명 개선

- H2ReservationRepository -> JdbcReservationRepository
- H2ReservationTimeRepository -> JdbcReservationTimeRepository

threepebbles committed:
refactor: ControllerAdvice -> RestControllerAdvice 수정

threepebbles committed:
refactor: 어노테이션 순서 배치를 일관되게 수정

- Spring Stereotype 어노테이션, 생성자/필드 관련 어노테이션, 핸들러 매핑 관련 어노테이션 순서로 정리

threepebbles committed:
refactor: 가독성 개선을 위해 메서드 체이닝 중간에 개행 추가

threepebbles committed:
docs: 기능 설명 추가

threepebbles committed:
refactor: view 이름 반환 시 불필요한 '/' 제거

threepebbles committed:
test: 빈 문자열 검증 테스트 추가

threepebbles committed:
test: 테마 생성, 삭제 예외 테스트 추가

threepebbles committed:
test: 변수명 변경

threepebbles committed:
chore: 예약 시간 스키마 변경

- start_at 타입: VARCHAR(255) -> TIME

threepebbles committed:
feat: 유저 예약 시스템에서 특정 날짜, 테마 id로 예약 가능한 시간을 조회하는 API의 HTTP 메서드 수정(POST -> GET)

threepebbles committed:
test: 예약 가능한 시간 목록 조회 API 테스트 개선

threepebbles committed:
refactor: toString 재정의에 Lombok 기능 사용

threepebbles commented:
안녕하세요 매트! 헤일러입니다.
주말에 약속이 있어서 피드백 반영이 다소 늦어진 점 죄송합니다 😅 
말씀해주신 피드백들 고민하고 반영해보았습니다.😀 
잘 부탁드립니다!

threepebbles commented:
안녕하세요 매트! 헤일러입니다.
주말에 약속이 있어서 피드백 반영이 다소 늦어진 점 죄송합니다 😅 
말씀해주신 피드백들 고민하고 반영해보았습니다.😀 
잘 부탁드립니다!

hyeonic commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,46 @@
+    private final String description;
+    private final String thumbnail;
+
+    public Theme(final Long id, final String name, final String description, final String thumbnail) {
+        validateName(name);
+        validateDescription(description);
+        validateThumbnail(thumbnail);
+        this.id = id;
+        this.name = name;
+        this.description = description;
+        this.thumbnail = thumbnail;
+    }
+
+    public Theme(final String name, final String description, final String thumbnail) {
+        this(null, name, description, thumbnail);
+    }
```
this 생성자를 활용하여 검증 로직을 잘 모아두었네요 👍

hyeonic commented on src/test/java/roomescape/reservation/service/ReservationTimeServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+import roomescape.reservation.repository.ReservationRepository;
+import roomescape.reservation.repository.ReservationTimeRepository;
+
+@JdbcTest
```
`@JdbcTest`는 기본적으로 어떠한 Bean들을 생성하나요?

hyeonic commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<Void> handleIllegalArgumentException(final IllegalArgumentException e) {
+        return ResponseEntity.badRequest().build();
+    }
+
+    @ExceptionHandler(ResourceNotFoundException.class)
+    public ResponseEntity<Void> handleDataNotFoundException(final ResourceNotFoundException e) {
+        return ResponseEntity.notFound().build();
+    }
+
+    @ExceptionHandler(AlreadyExistException.class)
+    public ResponseEntity<Void> handDataExistException(final AlreadyExistException e) {
+        return ResponseEntity.status(HttpStatus.CONFLICT).build();
+    }
+}
```
현재 해당 Handler 통해서 제어되는 예외는 `IllegalArgumentException`, `ResourceNotFoundException`, `AlreadyExistException`만 있는 것 같아요. 만약 핸들링 대상이 아닌 예외는 어떻게 될까요? 이를 막기 위해서는 어떻게 할 수 있을까요?

hyeonic commented on src/main/java/roomescape/reservation/repository/JdbcReservationRepository.java:
```diff
@@ -0,0 +1,154 @@
+        jdbcTemplate.update(sql, id);
+    }
+
+    @Override
+    public boolean existsByDateAndStartAtAndThemeId(final LocalDate date, final LocalTime startAt,
+                                                    final Long themeId) {
+        final String sql = """
+                SELECT EXISTS(
+                    SELECT 1
+                    FROM reservations AS r
+                    INNER JOIN reservation_times AS rt
+                    ON r.time_id = rt.id
+                    INNER JOIN themes AS th
+                    ON r.theme_id = th.id
+                    WHERE r.date = ? AND rt.start_at = ? AND th.id = ?
+                )
+                """;
+
+        return jdbcTemplate.queryForObject(sql, Boolean.class, date, startAt, themeId);
+    }
```
`jdbcTemplate.queryForObject(sql, Boolean.class, date, startAt, themeId)`을 수행하면 wrapper class `Boolean`을 반환할 것 같아요. existsByDateAndStartAtAndThemeId의 반환 타입과 일치시키거나 타입 캐스팅하면 좀 더 안전하게 활용 가능할 것 같습니다!

헤일러는 어떠한 상황에서 primitive type과 wrapper class를 고려하시나요?

hyeonic reviewed (approved):
안녕하세요 헤일러! 대부분 로직 잘 구현해주셔서 이후 단계 진행해도 좋을 것 같습니다.

추가적인 코멘트 남겨두었으니 이후 단계 진행할 때 같이 고려해주세요 ㅎㅎ 질문에 대한 답은 이후 PR이나 해당 PR에 편하신 곳에 남겨주시고 애매한 포인트가 있다면 언제든 편하게 DM 남겨주세요~
